### ISSUE TITLE
New lints: Safe items inside `unsafe extern`

### ISSUE THREAD

**Issue by obi1kenobi** [obi1kenobi/cargo-semver-checks#975]
A new feature in Rust 1.82+ is the ability to define safe items inside an `unsafe extern` block:
https://blog.rust-lang.org/2024/10/17/Rust-1.82.0.html#safe-items-with-unsafe-extern
```rust
unsafe extern {
    pub safe static TAU: f64;
    pub safe fn sqrt(x: f64) -> f64;
    pub unsafe fn strlen(p: *const u8) -> usize;
}
```

Making a previously-safe item become unsafe is a major breaking change, since it now requires `unsafe` at the point of use.

At least two lints, and possibly more:
- [x] safe extern static became unsafe
- [x] safe extern fn became unsafe

I believe we'll also need additional schema in [the adapter](https://github.com/obi1kenobi/trustfall-rustdoc-adapter) to model extern items and show which items are safe vs unsafe. This shouldn't be too difficult to add, though it's an extra step compared to "just writing a lint." For folks new to contributing to Trustfall, I'd recommend first writing another lint that doesn't need this extra step before moving to this issue.

**Comment by aDotInTheVoid** [comment]
rustdoc-json doesn't model functions declared in `extern` blocks differently to normal functions: https://github.com/rust-lang/rust/pull/131517. If you handle functions changing safety, then you handle extern functions as-well.

**Comment by obi1kenobi** [comment]
> rustdoc-json doesn't model functions declared in `extern` blocks differently to normal functions: [rust-lang/rust#131517](https://github.com/rust-lang/rust/pull/131517). If you handle functions changing safety, then you handle extern functions as-well.

Neat, TIL -- thank you!

We might still want a separate lint for the two cases so we can offer a more specific diagnostic message in each case. I'll make a test crate and see what happens currently, and go from there.

> Relatedly, you probably want a lint for changing `pub static X: i32 = 0;` to `pub static mut X: i32 = 0`, as the latter is unsafe to access.

Nice catch! I think it might already be on a to-do list somewhere, but I'll make sure of it :)

**Comment by obi1kenobi** [comment]
I decided the current lint for safe -> unsafe fn is fine here, and just added test cases here and in the adapter:
https://github.com/obi1kenobi/cargo-semver-checks/pull/1003
https://github.com/obi1kenobi/trustfall-rustdoc-adapter/pull/611

Leaving this issue open until we can lint for safe -> unsafe transitions in `extern static`.

Thanks again @aDotInTheVoid!

**Comment by aDotInTheVoid** [comment]
For the `extern static` case, you'll need the (not yet merged) https://github.com/rust-lang/rust/pull/133715 for rustdoc-json to have safety information for statics.

Also worth noting: `static`'s are safe by default, so going from `static X: i32` to `extern { static X: i32 }` is much more likely than `unsafe extern { safe static X: i32 }` to `unsafe extern { unsafe static X: i32 }`

**Comment by obi1kenobi** [comment]
Awesome, thanks for putting the PR together! I'll hold the next cargo-semver-checks release until that's merged and available in `rustdoc-types` so the new release can support the new format.

> Also worth noting: `static`'s are safe by default, so going from `static X: i32` to `extern { static X: i32 }` is much more likely than `unsafe extern { safe static X: i32 }` to `unsafe extern { unsafe static X: i32 }`

Oh good call! We definitely want to lint for that.

**Comment by Frank-III** [comment]
Since we have this in the rustdoc-apdapter https://github.com/obi1kenobi/trustfall-rustdoc-adapter/pull/725, we are ready for `safe extern static became unsafe` this lint? I would like to work on this!

**Comment by obi1kenobi** [comment]
Yes, and I believe I updated the adapter versions so you can use the newly-added field. Go for it!

**Comment by obi1kenobi** [comment]
The final missing part has been completed by #1090.

**Comment by aDotInTheVoid** [comment]
> > Relatedly, you probably want a lint for changing `pub static X: i32 = 0;` to `pub static mut X: i32 = 0`, as the latter is unsafe to access.
> 
> Nice catch! I think it might already be on a to-do list somewhere, but I'll make sure of it :)

I don't think this is tracked anywhere else yet. (At least I can't find it in #5 or open A-lint issues)

**Comment by Frank-III** [comment]
> > > Relatedly, you probably want a lint for changing `pub static X: i32 = 0;` to `pub static mut X: i32 = 0`, as the latter is unsafe to access.
> > 
> > 
> > Nice catch! I think it might already be on a to-do list somewhere, but I'll make sure of it :)
> 
> I don't think this is tracked anywhere else yet. (At least I can't find it in [#5](https://github.com/obi1kenobi/cargo-semver-checks/issues/5) or open A-lint issues)

I think it is added in  https://github.com/obi1kenobi/cargo-semver-checks/pull/1020

**Comment by obi1kenobi** [comment]
Yes, #1020 is where we merged the lint for that case — I just appear to have failed to update the issue tracker for it.
